x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java

```diff
@@ -176,12 +318,34 @@ public static User randomUser() {
         return new User(randomAlphaOfLengthBetween(3, 8), randomArray(1, 3, String[]::new, () -> randomAlphaOfLengthBetween(3, 8)));
     }
 
-    public static RealmRef randomRealm() {
-        return new RealmRef(
-            randomAlphaOfLengthBetween(3, 8),
-            randomFrom(FileRealmSettings.TYPE, NativeRealmSettings.TYPE, randomAlphaOfLengthBetween(3, 8)),
+    public static RealmRef randomRealmRef(boolean underDomain) {
+        final Supplier<String> randomRealmTypeSupplier = () -> randomFrom(
+            FileRealmSettings.TYPE,
+            NativeRealmSettings.TYPE,
+            LdapRealmSettings.AD_TYPE,
+            LdapRealmSettings.LDAP_TYPE,
+            JwtRealmSettings.TYPE,
+            OpenIdConnectRealmSettings.TYPE,
+            SamlRealmSettings.TYPE,
+            KerberosRealmSettings.TYPE,
             randomAlphaOfLengthBetween(3, 8)
         );
+        RealmDomain domain = null;
+        if (underDomain) {
+            final Set<RealmConfig.RealmIdentifier> domainRealms = Set.of(
+                randomArray(
+                    1,
+                    4,
+                    RealmConfig.RealmIdentifier[]::new,
+                    () -> new RealmConfig.RealmIdentifier(
+                        randomRealmTypeSupplier.get(),
+                        randomAlphaOfLengthBetween(3, 8).toLowerCase(Locale.ROOT)
+                    )
+                )
+            );
+            domain = new RealmDomain("domain", domainRealms);
+        }
+        return new RealmRef(randomAlphaOfLengthBetween(3, 8), randomRealmTypeSupplier.get(), randomAlphaOfLengthBetween(3, 8), domain);
```

Comment:
Yes, that's correct.
The `RealmDomain` should be mocked in these tests but this is not possible because `record`s are final, which can't be mocked, so I recurred to some realm domains...
Like we agreed I pushed [3c7be85](https://github.com/elastic/elasticsearch/pull/82639/commits/3c7be8557b78479a0c89f06f5311a8dc806a1e81) in order to build a correct `RealmRef`.